Feature/use cmake init

ClausKlein opened the pull request.

ClausKlein commented on asio/cmake/install-config.cmake:
```diff
@@ -0,0 +1,4 @@
+include(CMakeFindDependencyMacro)
+find_dependency(Threads)
```
@[friendlyanon](https://github.com/friendlyanon) needed to add this to make the config package useable for clients

friendlyanon replied:
Yes, this is covered by the templates generated using a package manager and this example: https://github.com/friendlyanon/cmake-init-transitive

ClausKlein commented on asio/cmake/install-rules.cmake:
```diff
@@ -0,0 +1,38 @@
+if(PROJECT_IS_TOP_LEVEL)
+  set(CMAKE_INSTALL_INCLUDEDIR include CACHE PATH "")
```
@[friendlyanon](https://github.com/friendlyanon) This was wrong `include/asio` generated and was in wrong scope!

friendlyanon replied:
This is explained by the pinned issue and the comments within: https://github.com/friendlyanon/cmake-init/issues/43

If you don't wish to protect the user from that by default, then this bit can be removed entirely.

ClausKlein commented on asio/CMakeLists.txt:
```diff
@@ -0,0 +1,51 @@
+target_include_directories(
+  asio SYSTEM # XXX ${warning_guard}
+  INTERFACE "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>" #
+            "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>"
```
@[friendlyanon](https://github.com/friendlyanon) I missed this `SYSTEM` include!

friendlyanon replied:
`INSTALL_INTERFACE` is unnecessary here. The `install(TARGETS)`'s special form (`INCLUDE DESTINATION`) already adds the include path to the install interface.

`SYSTEM` is already added with the `warning_guard` variable when necessary. `warning_guard` is for vendoring projects, see the comment in `variables.cmake`.

ClausKlein commented on asio/CMakeLists.txt:
```diff
@@ -0,0 +1,51 @@
+include(cmake/variables.cmake)
+
+if(PROJECT_IS_TOP_LEVEL)
+  include(GNUInstallDirs) # for CMAKE_INSTALL_INCLUDEDIR
```
@[friendlyanon](https://github.com/friendlyanon) This `include` is needed in this context!

friendlyanon replied:
This is not needed here, the code that deals with installing is in the `install-rules.cmake` file.

ClausKlein replied:
but not if `CMAKE_SKIP_INSTALL_RULES` is set!

ClausKlein commented on asio/.cmake-format:
```diff
@@ -0,0 +1,30 @@
+  dangle_parens: true
+  keyword_case: upper
+  line_ending: unix
+  line_width: 120
```
I formatted all `cmake files` this this configuration!

[friendlyanon](https://github.com/friendlyanon) **TODO**: `cmake-linit cmake/*.cmake`

friendlyanon replied:
Sure, I don't mind. Personally I don't use this tool, because it results in weird formatting.

friendlyanon commented on asio/CMakeUserPresets.json:
```diff
@@ -0,0 +1,35 @@
+{
```
This file should NOT be in version control: https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html

ClausKlein replied:
I know, but the file get always lost with `git clean -xdf`

friendlyanon commented on asio/cmake/WarningsAsErrors.cmake:
```diff
@@ -0,0 +1,12 @@
+if(CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND CMAKE_CXX_SIMULATE_ID MATCHES "MSVC")
```
This code should be in the user presets file. Definitely not hardcoded in the CMake code.

Warnings as errors is also counterproductive:  
https://youtu.be/bsXLMQ6WgIk?t=3912  
https://youtu.be/bXx3-drA_oI

ClausKlein replied:
yes, but they are  `PRIVATE`, and not exported.

friendlyanon replied:
It's still a lot of code for something that can be a short, single line in the presets. Developer mode is also erroneously on by default, which would affect builds when it shouldn't.

ClausKlein replied:
not longer need

friendlyanon commented on asio/cmake/dev-mode.cmake:
```diff
@@ -0,0 +1,27 @@
+include(cmake/folders.cmake)
+
+#XXX include(CTest)
+option(BUILD_TESTING "Build the testing tree." ON)
```
I'm curious what the reason for this is.

ClausKlein replied:
The `CTest` module creates lots of targets you only need if you are working with [CDash](https://cmake.org/cmake/help/book/mastering-cmake/chapter/CDash.html)

friendlyanon replied:
Sure. A short comment above the option explaining that would be neat then.

ClausKlein replied:
Yes, good idea!

Thank you for the review

friendlyanon commented on asio/cmake/variables.cmake:
```diff
@@ -0,0 +1,24 @@
+# Targets necessary to build the project must be provided unconditionally, so
+# consumers can trivially build and package the project
+if(PROJECT_IS_TOP_LEVEL)
+  option(asio_DEVELOPER_MODE "Enable developer mode" ON)
```
This must be `OFF` by default. Users must not be bothered by this project's developer mode.

ClausKlein replied:
YES, but the should use it with `CMP.cmake `as a subproject.

friendlyanon reviewed (changes requested):
It looks like your mentions somehow got messed up, as they are not really mentions.

I commented on some of the things, but the one thing that is of utmost importance is to not have the developer mode be enabled by default. The reason for this is explained in the HACKING document. Contrast that with the BUILDING document, which is the user facing document and is assumed to be the default case.

ClausKlein commented on asio/CMakeLists.txt:
```diff
@@ -14,14 +14,21 @@ include(cmake/variables.cmake)
 
 if(PROJECT_IS_TOP_LEVEL)
   include(GNUInstallDirs) # for CMAKE_INSTALL_INCLUDEDIR
+  set(CMAKE_VERIFY_INTERFACE_HEADER_SETS ON)
```
@friendlyanon Do you know this new **CMake** feature?

It works together with 
```cmake
target_sources(asio INTERFACE FILE_SET HEADERS BASE_DIRS include FILES ${headers})
```

see too [target_sources](https://cmake.org/cmake/help/v3.24/manual/cmake-buildsystem.7.html?highlight=target_sources#interface-libraries)

ClausKlein commented on asio/cmake/install-rules.cmake:
```diff
@@ -13,7 +13,9 @@ set(package asio)
 install(DIRECTORY include/ DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}" COMPONENT asio_Development)
 
-install(TARGETS asio EXPORT asioTargets INCLUDES DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
+install(TARGETS asio EXPORT asioTargets INCLUDES DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
+        FILE_SET HEADERS
+)
```
@friendlyanon  `CMAKE_VERIFY_INTERFACE_HEADER_SETS` requires this too!

ClausKlein committed:
merge in cmake-init generated files

asio needs Threads library, so this is added to install-config.cmake

export ASIO_NO_DEPRECATED definition

add cmake user preset too

do not convert british to american english

add todos from codespell check

liberate clang-tidy config file for use with async code too

use cmake VERIFY_INTERFACE_HEADER_SETS

install pkg config file too

suppress compiler Warnings with llvm-17 on OSX

do not verify nor install experimental headers

disalbe some warnings in developer mode

ClausKlein committed:
Update cmake files with cmake-init v0.40.5

Add requirements.txt for pip

Add ignore-workd-list for codespell

Add CI cmake based project workflow

ClausKlein committed:
Add CMake Workflow Presets

Update used tools versions on CI build

ClausKlein committed:
Use clang-tidy with CI sanitize preset

ClausKlein committed:
Suppress more clang-17 warnings

ClausKlein committed:
Add ASIO_SEPARATE_COMPILATION option

ClausKlein committed:
Check only public interface header, no details

ClausKlein committed:
Use hostSystemName for presets

ClausKlein committed:
Update GNUmakefile and *.md files

ClausKlein committed:
Fix some clang-tidy warnings

ClausKlein committed:
Modernize GNUmakefile

ClausKlein committed:
Disable openssl files for now

ClausKlein committed:
update

ClausKlein commented:
Old asio source code version not longer supported!
